prov/verbs/src/verbs_info.c

```diff
@@ -199,19 +199,22 @@ int fi_ibv_check_rx_attr(const struct fi_rx_attr *attr,
 			 const struct fi_info *hints,
 			 const struct fi_info *info)
 {
-	/* WARNING: This is not thread safe */
-	uint64_t saved_prov_mode = info->rx_attr->mode;
+	struct fi_info *dup_info;
 	int ret;
 
-	info->rx_attr->mode = (hints->domain_attr &&
-			       hints->domain_attr->cq_data_size) ?
-			info->rx_attr->mode :
-			(info->rx_attr->mode & ~FI_RX_CQ_DATA);
-
-	ret = ofi_check_rx_attr(&fi_ibv_prov, info, attr, hints->mode);
-
-	info->rx_attr->mode = saved_prov_mode;
+	if (hints->mode & FI_RX_CQ_DATA || (hints->rx_attr &&
+	    (hints->rx_attr->mode & FI_RX_CQ_DATA))) {
+		ret = ofi_check_rx_attr(&fi_ibv_prov, info, attr, hints->mode);
+	} else {
+		dup_info = fi_dupinfo(info);
+		if (!dup_info)
+			return -FI_ENOMEM;
 
+		dup_info->rx_attr->mode &= ~FI_RX_CQ_DATA;
+		ret = ofi_check_rx_attr(&fi_ibv_prov, dup_info, attr,
+					hints->mode);
+		fi_freeinfo(dup_info);
```

Comment:
I think we need to have the check for `domain_attr->cq_data_size`.

> It is an error to set domain->cq_data_size and not also
include FI_RX_CQ_DATA.

The only way to detect that is by not removing `FI_RX_CQ_DATA` from `info` when `domain_attr->cq_data_size` is set. Then `ofi_check_rx_attr` would fail if app hasn't set the mode bit in `hints`

Reply:
Yes, that makes sense; I'll update the PR. Thanks.

Reply:
Updated the PR to make non-zero cq_data_size validate for FI_RX_CQ_DATA, and updated the PR commit message.